build_tools/docker/manage_images.py

```diff
@@ -130,144 +131,148 @@ def parse_arguments():
+  return processing_order
 
 
-def stream_command(command, dry_run=False):
+def run_command(command: Sequence[str],
+                dry_run: bool = False,
+                check: bool = True,
+                capture_output: bool = False,
+                universal_newlines: bool = True,
+                **run_kwargs) -> subprocess.CompletedProcess:
+  """Thin wrapper around subprocess.run"""
   print(f'Running: `{" ".join(command)}`')
-  if dry_run:
-    return 0
-  process = subprocess.Popen(command,
-                             bufsize=1,
-                             stderr=subprocess.STDOUT,
-                             stdout=subprocess.PIPE,
-                             universal_newlines=True)
-  for line in process.stdout:
-    print(line, end='')
-
-  if process.poll() is None:
-    raise RuntimeError('Unexpected end of output while process is not finished')
-  return process.poll()
-
-
-def check_stream_command(command, dry_run=False):
-  exit_code = stream_command(command, dry_run=dry_run)
-  if exit_code != 0:
-    print(f'Command failed with exit code {exit_code}: `{" ".join(command)}`')
-    sys.exit(exit_code)
-
-
-def get_repo_digest(image):
+  if not dry_run:
+    if capture_output:
+      # Hardcode support for python <= 3.6.
+      run_kwargs['stdout'] = subprocess.PIPE
+      run_kwargs['stderr'] = subprocess.PIPE
+
+    completed_process = subprocess.run(command,
+                                       universal_newlines=universal_newlines,
+                                       check=check,
+                                       **run_kwargs)
+    return completed_process
+  # Dummy CompletedProess with successful returncode.
+  return subprocess.CompletedProcess(command, returncode=0)
+
+
+def get_repo_digest(tagged_image_url: str) -> str:
   inspect_command = [
       'docker',
       'image',
       'inspect',
-      f'{image}',
+      tagged_image_url,
       '-f',
       '{{index .RepoDigests 0}}',
   ]
-  inspect_process = subprocess.run(inspect_command,
-                                   universal_newlines=True,
-                                   stdout=subprocess.PIPE,
-                                   stderr=subprocess.PIPE,
-                                   timeout=10)
-  if inspect_process.returncode != 0:
-    print(f'Computing the repository digest for {image} failed.'
+  try:
+    completed_process = run_command(
+        inspect_command,
+        dry_run=False,  # Run even if --dry_run is True.
+        capture_output=True,
+        timeout=10)
+  except subprocess.CalledProcessError as error:
+    print(f'Computing the repository digest for {tagged_image_url} failed.'
           ' Has it been pushed to GCR?')
-    print(f'Output from `{" ".join(inspect_command)}`:')
-    print(inspect_process.stdout, end='')
-    print(inspect_process.stderr, end='')
-    sys.exit(inspect_process.returncode)
-  _, repo_digest = inspect_process.stdout.strip().split('@')
+    raise error
```

Comment:
Rather than print, what about using "raise from" https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement

of course GitHub is being trash and not letting me write a suggestion, but:

```suggestion
  raise RuntimeError(f'Computing the repository digest for {tagged_image_url}'
                     ' failed. Has it been pushed to GCR?') from error
```

Same below

Reply:
Cool :)